[Doc] Subqueries in expand (#497)
SAP/cloud-sdk: docs/js/features/odata/odata-v4-client.mdx

```sql
$filter=(/any(a0:((a0/Friends/FirstName eq 'firstName' and a0/Friends/LastName eq 'lastName'))))
```

### Expand

#### Expand and Select

In contrast to the OData v2 implementation, you have `select` and `expand` separately.
In other words selected properties are not expanded automatically as in v2.

```ts
People.requestBuilder().getAll()
    .select(People.LAST_NAME)
    .expand(People.FRIENDS.select(People.FIRST_NAME,People.ADDRESS_INFO))
```

In the example above you select the `LAST_NAME` of the root entity and expand the navigation property `FRIENDS`.
In the expanded entity the selected fields are `FIRST_NAME` and `ADDRESS_INFO`.

The generated URL for this request will be:
```sql
/People?$select=LastName&$expand=Friends($select=FirstName,AddressInfo)
```
If no select is given, all non-navigational properties are included in the response.

#### Sub-queries in Expand

> Note that you can crate very complex queries within the expand scope:

Reviewer: ```suggestion
Note that you can create very complex queries within the expand scope:
```

```sql
$filter=(/any(a0:((a0/Friends/FirstName eq 'firstName' and a0/Friends/LastName eq 'lastName'))))
```

### Expand

#### Expand and Select

> In contrast to the OData v2 implementation you have select and expand separately.

Reviewer: ```suggestion
In contrast to the OData v2 implementation, you have `select` and `expand` separately.
```

In other words selected properties are not expanded automatically as in v2.

```ts
People.requestBuilder().getAll()
    .select(People.LAST_NAME)
    .expand(People.FRIENDS.select(People.FIRST_NAME,People.ADDRESS_INFO))
```

In the example above you select the `LAST_NAME` of the root entity and expand the navigation property `FRIENDS`.
In the expanded entity the selected fields are `FIRST_NAME` and `ADDRESS_INFO`.

The generated URL for this request will be:
```sql
/People?$select=LastName&$expand=Friends($select=FirstName,AddressInfo)
```

> If no select is given all non-navigational properties are included in the response.

Reviewer: ```suggestion
If no select is given, all non-navigational properties are included in the response.
```

#### Sub-queries in Expand

Note that you can crate very complex queries within the expand scope:
```ts
People.requestBuilder().getAll()
    .expand(People.FRIENDS
      .select(People.FIRST_NAME,People.ADDRESS_INFO)
      .filter(People.LAST_NAME.equals("Miller"))
      .orderBy(asc(People.GENDER))
      .top(1)
      .skip(1)
    )
```

In this example the filter will reduce the friends to be shown.
The effect of a filter depends whether it is used in or outside of an expand.
The different cases are explained in [Filters in Expand.](#filterParentChildren)

The URL for the query will be:
```sql

> /People?$expand=Friends($select=FirstName,AddressInfo;$filter=(LastName eq 'Miller');$skip=1;$top=1)

Reviewer: the order by is missing

Author: good catch. I think we had a bug there. I just copied it from the debug console. https://github.com/SAP/cloud-sdk/pull/504

```

#### Filter parent vs. Filter children <a name="filterParentChildren"/>

Depending on the context of the filter it will either filter the parent or the children.
In our example we have a `PERSON` related to 0..N `FRIENDS` which are both of type people.

> If you want to get all persons with first name `John` the query is:

Reviewer: ```suggestion
If you want to get all people with first name `John` the query is:
```

```ts
People.requestBuilder().getAll()
    .filter(People.FIRST_NAME.equals('John'))
```

> If you want to get all persons who have at least one friend with first name `John` the query is:

Reviewer: ```suggestion
If you want to get all people who have at least one friend with first name `John` the query is:
```

```ts
People.requestBuilder().getAll().filter(People.FRIENDS
    .filter(
        any(People.FIRST_NAME.equals('John'))
    )
)
```

> The lambda `all` would enforce that all friends must have first name `John`.

Reviewer: for the lambda expression `all`, maybe a link to the api doc is helpful

Author: yes

The two queries above filter the parent entity person.

> In case you want to get all persons, but reduce the friends in the response the filter has to be inside the expand:

Reviewer: ```suggestion
In case you want to get all people but reduce the friends in the response, the filter has to be inside the expand:
```

```ts
People.requestBuilder().getAll()
    .expand(People.FRIENDS
        .filter(People.FIRST_NAME.equals('John'))
    )
```

> This will return all persons but only the friends with name `John` will be included in the response.

Reviewer: ```suggestion
This will return all people but only the friends with first name `John` will be included in the response.
```

> 
> 
> 
> 
> 
> 
> 
> 
> 

Reviewer: ```suggestion
```
